src/builtins/path.cpp

```diff
@@ -0,0 +1,895 @@
+        if (*opts.arg1) {
+            if (opts.arg1[0] != L'.') {
+                ext.push_back(L'.');
+            }
+            ext.append(opts.arg1);
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_resolve(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
+            // something that does.
+            wcstring next = *arg;
+            // First add $PWD if we're relative
+            if (!next.empty() && next[0] != L'/') {
+                // Note pwd can have symlinks, but we are about to resolve it anyway.
+                next = path_apply_working_directory(*arg, parser.vars().get_pwd_slash());
+            }
+            auto rest = wbasename(next);
+            while(!next.empty() && next != L"/") {
+                next = wdirname(next);
+                real = wrealpath(next);
+                if (real) {
+                    real->push_back(L'/');
+                    real->append(rest);
+                    real = normalize_path(*real, false);
+                    break;
+                }
+                rest = wbasename(next) + L'/' + rest;
+            }
+            if (!real) {
+                continue;
+            }
+        }
+
+        // Normalize the path so "../" components are eliminated even after
+        // nonexistent or non-directory components.
+        // Otherwise `path resolve foo/../` will be `$PWD/foo/../` if foo is a file.
+        real = normalize_path(*real, false);
+
+        // Return 0 if we found a realpath.
+        if (opts.quiet) {
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, *real);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_sort(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    opts.invert_valid = true;
+    opts.key_valid = true;
+    opts.unique_valid = true;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    auto func = +[] (const wcstring &x) {
+        return wbasename(x);
+    };
+    if (opts.have_key) {
+        if (std::wcscmp(opts.key, L"basename") == 0) {
+            // Do nothing, this is the default
+        } else if (std::wcscmp(opts.key, L"dirname") == 0) {
+            func = +[] (const wcstring &x) {
```

Comment:
I guess we don't need the + here. Also maybe call it something like `keyfunc`?

Reply:
Renamed to keyfunc (and "funced" to "key", because that's what it does - it gets the key to a value).

I don't want to drop the "+" here because that leaves us with a weird asymmetry.

Reply:
> I don't want to drop the "+" here because that leaves us with a weird asymmetry.

Sure. IMO the asymmetry is justified because `+` ist just a short way of spelling a declaration's type.
When I first tried to figure out what the `+` means I removed it and recompiled, but this one actually makes no difference which makes Occam's razor less sharp.